Add support for inline_asm_elementwise in Inductor lowerings

Chillee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #129950
* __->__ #129846



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

This doesn't actually expose `inline_asm_elementwise` from any public API, but makes it pretty easy to register a lowering for a custom op that uses it.

<img width="667" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/f125f4bb-4f8c-46e7-8e06-925f37ed2930">

Chillee committed:
Add support for inline_asm_elementwise in Inductor lowerings

[ghstack-poisoned]

yanboliang commented on test/inductor/test_custom_lowering.py:
```diff
@@ -97,6 +99,45 @@ def inner_fn(index):
+
+        register_lowering(
+            torch.ops.test_inductor_ops.tanh_approx, type_promotion_kind=None
+        )(tanh_approx_lowering)
```
You want to register ```add_custom_lowering````?

Chillee committed:
Update on "Add support for inline_asm_elementwise in Inductor lowerings"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

This doesn't actually expose `inline_asm_elementwise` from any public API, but makes it pretty easy to register a lowering for a custom op that uses it.

<img width="667" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/f125f4bb-4f8c-46e7-8e06-925f37ed2930">

[ghstack-poisoned]

shunting314 commented on torch/_inductor/codegen/triton.py:
```diff
@@ -721,6 +721,13 @@ def maximum(a, b):
+        triton_type = triton_compute_type(dtype)
+        input_refs = ", ".join([str(i) for i in inputs])
+        constraints = ", ".join(["=r"] + ["r" for _ in inputs])
+        return f"tl.inline_asm_elementwise('{asm}', '{constraints}', [{input_refs}], dtype={triton_type}, is_pure=True, pack=1)"
```
Do we want to expose pack as an parameter to `inline_asm_elementwise`?

eellison commented on torch/_inductor/codegen/triton.py:
```diff
@@ -721,6 +721,13 @@ def maximum(a, b):
         return f"tl.where({a}, {b}, {c})"
 
+    @staticmethod
+    def inline_asm_elementwise(*inputs, asm, dtype=torch.float32):
```
What about all the other arguments:

```
Parameters
:
asm – assembly to run. Must match target’s assembly format.

constraints – asm constraints in [LLVM format](https://llvm.org/docs/LangRef.html#inline-asm-constraint-string)

args – the input tensors, whose values are passed to the asm block

dtype – the element type(s) of the returned tensor(s)

is_pure – if true, the compiler assumes the asm block has no side-effects

pack – the number of elements to be processed by one instance of inline assembly

_builder – the builder
```

Would it make sense to expose any of them in api ?

Chillee replied:
To be clear, this is currently not a "public" API - you can only access it by registering a custom lowering and calling `ops`. In principle, some of these might make sense, but I don't really understand what most of them would be used for, so I only expose `asm` and `dtype` for now.

Chillee replied:
Maybe I should just expose them in-case somebody wants to use them 🤔

shunting314 reviewed (approved):
What are the general uses cases of using inline assembly like this?

Is this to cover the cases where triton does not have a corresponding op in its TTIR?

drisspg commented on torch/_inductor/kernel/flex_attention.py:
```diff
@@ -547,8 +547,8 @@ def flex_attention(*args, **kwargs):
             # For now, we always assume the "sound" option
-            SCORE_MOD_IS_LINEAR=False,
-            ROWS_GUARANTEED_SAFE=False,
+            SCORE_MOD_IS_LINEAR=True,
```
dont land this .. haha

Chillee committed:
Update on "Add support for inline_asm_elementwise in Inductor lowerings"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

This doesn't actually expose `inline_asm_elementwise` from any public API, but makes it pretty easy to register a lowering for a custom op that uses it.

<img width="667" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/f125f4bb-4f8c-46e7-8e06-925f37ed2930">

[ghstack-poisoned]

Chillee commented:
@shunting314 Yeah, there a variety of `ptx` instructions that may not be explicitly exposed to the user. For example, in this case, `libdevice.tanh` gets lowered to this in SASS

![image](https://github.com/pytorch/pytorch/assets/6355099/80a7e310-83d4-4255-acb9-dc7bfc886cb4)

And this causes a 2.5x slowdown in FlashAttention.

OTOH, if you use `tanh.approx`, there's only about a 15% slowdown.

shunting314 commented:
> OTOH, if you use tanh.approx, there's only about a 15% slowdown.

BTW, tanh.approx returns an approximation? Does that hurt accuracy or it's negligible?

Chillee commented:
@shunting314 Yeah, it definitely hurts accuracy. But in cases like this, users might want to trade off a 2.5x perf improvement for degraded accuracy.

Chillee committed:
Update on "Add support for inline_asm_elementwise in Inductor lowerings"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

This doesn't actually expose `inline_asm_elementwise` from any public API, but makes it pretty easy to register a lowering for a custom op that uses it.

<img width="667" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/f125f4bb-4f8c-46e7-8e06-925f37ed2930">

[ghstack-poisoned]

Chillee committed:
Update on "Add support for inline_asm_elementwise in Inductor lowerings"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

This doesn't actually expose `inline_asm_elementwise` from any public API, but makes it pretty easy to register a lowering for a custom op that uses it.

<img width="667" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/f125f4bb-4f8c-46e7-8e06-925f37ed2930">

[ghstack-poisoned]

Chillee commented:
@pytorchbot merge -i

Chillee commented:
@pytorchbot merge -i

Chillee commented:
/easycla

Chillee commented:
@pytorchbot merge -i

Chillee commented:
@pytorchbot merge -i

jeffdaily commented:
@pytorchbot revert -m "introduced ROCm breakages in trunk" -c nosignal

jeffdaily commented:
Reverted because ROCm does not support CUDA-specific inline asm.  If it were a single unit test that was added by this PR, we would have opened an issue to disable the test.  We try to avoid opening multiple DISABLED issues when they can be root-caused back to a single PR like this.  This PR introduces 2 unit tests.  I can understand you might not be able to modify these tests to also work correctly for ROCm, so in this case please add `@skipCUDAIfRocm` decorator to these tests.  I have also added the ciflow/rocm label to run the full ROCm CI which caught this after the merge.

Chillee commented:
Oh oops, should have thought about that.

Chillee committed:
Update on "Add support for inline_asm_elementwise in Inductor lowerings"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

This doesn't actually expose `inline_asm_elementwise` from any public API, but makes it pretty easy to register a lowering for a custom op that uses it.

<img width="667" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/f125f4bb-4f8c-46e7-8e06-925f37ed2930">

[ghstack-poisoned]

Chillee committed:
Update on "Add support for inline_asm_elementwise in Inductor lowerings"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

This doesn't actually expose `inline_asm_elementwise` from any public API, but makes it pretty easy to register a lowering for a custom op that uses it.

<img width="667" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/f125f4bb-4f8c-46e7-8e06-925f37ed2930">

[ghstack-poisoned]

Chillee commented:
@pytorchbot merge -i

Chillee commented:
@pytorchbot merge -i
